### PR TITLE
[SYCL] Address multi_ptr intel/llvm vs spec mismatches.

### PR DESCRIPTION
**Pull request by mmoadeli**
* Deprecate the multi_ptr constructor from accessor(Reg, Void) with target::local.
* Deprecate the multi_ptr constructor from accessor(Void) and local_accessor with generic_space.
* Deprecate the multi_ptr constructor from Legacy using local_accessor.
* Add const qualification to the explicit conversion of multi_ptr<generic_space> to multi_ptr of any other address space.
* Adjust the order of template parameter deduction guides in sync with the specification.
* Implement support for the constant_space specialization of non-legacy multi_ptr.

### CODE REVIEW

**Comment by againull** [comment]
@mmoadeli Could you please fix LIT failures.

File: sycl/include/sycl/multi_ptr.hpp [hunk 1]
```diff
@@ -148,6 +148,8 @@ class multi_ptr {
                 RelaySpace == Space &&
                 (Space == access::address_space::generic_space ||
                  Space == access::address_space::local_space)>>
+  __SYCL2020_DEPRECATED("multi_ptr construction using target::local spcialized "
```
Comment (againull) [inline sycl/include/sycl/multi_ptr.hpp:151]:
```suggestion
  __SYCL2020_DEPRECATED("multi_ptr construction using target::local specialized "
```

File: sycl/include/sycl/multi_ptr.hpp [hunk 2]
```diff
@@ -464,17 +470,20 @@ class multi_ptr<const void, Space, DecorateAddress> {
       access::address_space RelaySpace = Space,
       typename = typename std::enable_if_t<
           RelaySpace == Space && Space == access::address_space::local_space>>
+  __SYCL2020_DEPRECATED("multi_ptr construction using target::local spcialized "
```
Comment (againull) [inline sycl/include/sycl/multi_ptr.hpp:473]:
```suggestion
  __SYCL2020_DEPRECATED("multi_ptr construction using target::local specialized "
```

**Comment by againull** [comment]
Could you please merge origin/sycl branch because some of the tests failed because of misalignment between test-e2e and compiler (see https://github.com/intel/llvm/pull/10334 for details)

**Comment by againull** [comment]
@mmoadeli Could you please the following failure:
Failed Tests (1):
SYCL :: multi_ptr/ctad.cpp

**Comment by againull** [comment]
> @mmoadeli Could you please fix the following failure: Failed Tests (1): SYCL :: multi_ptr/ctad.cpp

This failure is caused by this commit: https://github.com/intel/llvm/pull/10297/commits/d3f31a22fc3ad5c572b28d36372d5025033af403

Could you please point me to the specification where I can see this? Maybe I'm looking to the wrong place, but in this spec in 4.7.7.2. Explicit pointer aliases section I see that decoration is legacy:

```
// Deprecated in SYCL 2020
template <typename ElementType>
using constant_ptr =
    multi_ptr<ElementType, access::address_space::constant_space,
              access::decorated::legacy>;
```

**Comment by againull** [comment]
> > @mmoadeli Could you please fix the following failure: Failed Tests (1): SYCL :: multi_ptr/ctad.cpp
> 
> This failure is caused by this commit: [d3f31a2](https://github.com/intel/llvm/commit/d3f31a22fc3ad5c572b28d36372d5025033af403)
> 
> Could you please point me to the specification where I can see this deduction rule?
It looks before this change legacy specialization of multi_ptr was picked. But after this change (decoration changed from legacy to "no") SYCL 2020 multi_ptr is picked and assert is thrown because SYCL 2020 multi_ptr doesn't allow constant_buffer.

**Comment by mmoadeli** [comment]
> [againull](/againull)

**Comment by mmoadeli** [comment]
>

**Comment by mmoadeli** [comment]
> > > @mmoadeli Could you please fix the following failure: Failed Tests (1): SYCL :: multi_ptr/ctad.cpp
> > 
> > 
> > This failure is caused by this commit: [d3f31a2](https://github.com/intel/llvm/commit/d3f31a22fc3ad5c572b28d36372d5025033af403)
> 
> Could you please point me to the specification where I can see this deduction rule? It looks before this change legacy specialization of multi_ptr was picked. But after this change (decoration changed from legacy to "no") SYCL 2020 multi_ptr is picked and assert is thrown because SYCL 2020 multi_ptr doesn't allow constant_buffer.

@againull This change is not still part of spec. There is a pending PRhttps://github.com/KhronosGroup/SYCL-Docs/pull/439  for it

**Review by againull — APPROVED** [review]

**Pull request merged**